test/Kestrel.Core.Tests/Http2ConnectionTests.cs

```diff
@@ -3324,6 +3324,54 @@ public async Task CompletionSendsFinalGOAWAY()
             VerifyGoAway(await ReceiveFrameAsync(), 0, Http2ErrorCode.NO_ERROR);
         }
 
+        [Fact]
+        public async Task StopProcessingNextRequestSendsGracefulGOAWAYAndWaitsForStreamsToComplete()
+        {
+            var requestSemaphore = new SemaphoreSlim(1);
+            await InitializeConnectionAsync(context =>
+            {
+                requestSemaphore.Wait();
+                return Task.CompletedTask;
+            });
+
+            // Send and receive an unblocked request
+            await StartStreamAsync(1, _browserRequestHeaders, endStream: true);
+
+            await ExpectAsync(Http2FrameType.HEADERS,
+                withLength: 55,
+                withFlags: (byte)Http2HeadersFrameFlags.END_HEADERS,
+                withStreamId: 1);
+            await ExpectAsync(Http2FrameType.DATA,
+                withLength: 0,
+                withFlags: (byte)Http2DataFrameFlags.END_STREAM,
+                withStreamId: 1);
+
+            // Send a blocked request
+            await StartStreamAsync(3, _browserRequestHeaders, endStream: false);
+
+            // Close pipe
+            _pair.Application.Output.Complete();
+
+            // Assert connection closed
+            await _closedStateReached.Task.DefaultTimeout();
+            VerifyGoAway(await ReceiveFrameAsync(), 3, Http2ErrorCode.NO_ERROR);
+
+            // Assert connection shutdown is still blocked
+            // ProcessRequestsAsync completes the connection's Input pipe
+            var readTask = _pair.Application.Input.ReadAsync();
+            _pair.Application.Input.CancelPendingRead();
+            var result = await readTask;
+            Assert.False(result.IsCompleted);
+
+            // Unblock the request and ProcessRequestsAsync
+            requestSemaphore.Release();
+            await _connectionTask;
+
+            // Assert connection's Input pipe is completed
+            result = await _pair.Application.Input.ReadAsync();
```

Comment:
Nit: .DefaultTimeout()

Reply:
Also, for consistency, use the same pattern you used above of calling ReadAsync, then CancelPendingRead, then awaiting the task. It should return a completed task before CancelPendingRead is even called if you use a TCS without the RunContinuationsAsync option. This is the one of the few times I would recommend not using that option as long as you add a comment explaining why.

Reply:
You can't use .DefaultTimeout() here since ReadAsync() returns a ValueTask

Reply:
You could always call AsTask() first, but there shouldn't be much risk of a hanging test even if this regresses as long as the extra CancelPendingRead() is added anyway.